Schedule when all servers are ready + single mgr command for all servers + misc enhancements

suresh-thelkar opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
<!--- Describe the problem, ideally from the customer's viewpoint  -->
This PR consists of the following enhancements.

- qmgr/pbs_manager request should go to all servers automatically.
- Even if one of the configured servers is not available qmgr update should fail. But the stats queries should be successful when queried through external clients. But when they are called from scheduler the operation should fail.
- Server should stop contacting Scheduler when scheduling is off. This helps in not logging the following message to server logs continuously thus by filling up all the system when scheduler is gone down.
        08/25/2020 08:16:29;0001;Server@spark1;Svr;Server@spark1;Operation now in progress (115) in 
        contact_sched, Could not contact Scheduler
        08/25/2020 08:16:29;0001;Server@spark1;Svr;Server@spark1;Operation now in progress (115) in 
        contact_sched, Could not contact Scheduler
- Even default scheduler's host and port should be allowed to change by Admins. This is because only scheduler runs in Multi-Server and all other Servers should talk to the same scheduler.
- sched_host/sched_port if changed should remain persisted after server restarts. This is because Scheduler always overwrites them as part of the attribute updates when Server first contacts Scheduler. Modified this part of the code so that the changes are retained after server/s restarts.
- Similarly we should allow admin's to change sched_priv/sched_logs for default directory. This is again for the reasons mentioned above
- In fact following is recommended for Admin as a one time activity during the MultiServer setup. Set all of these values correctly so that remaining servers will connect to the same scheduler.
    qmgr -c "s sc sched_host=spark1,sched_port=15004"
    qmgr -c "s sched sched_log=/var/spool/pbs/sched_logs"
    qmgr -c "s sched sched_priv=/var/spool/pbs/sched_priv"
- Scheduling does not happen if one server goes down
  - untill all servers connected
  - if any server goes down while scheduler queries
#### Describe Your Change
<!--- Say how you fixed the problem.  Please describe your code changes in detail for reviewer -->


#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->



<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

nithinj commented on src/lib/Libifl/pbsD_statsched.c:
```diff
@@ -67,5 +67,58 @@
 __pbs_statsched(int c, struct attrl *attrib, char *extend)
 {
-	return PBSD_status_random(c, PBS_BATCH_StatusSched, "", attrib, extend, MGR_OBJ_SCHED);
+	struct batch_status *ret = NULL;
```
Why this is changed instead of changing PBSD_status_random? - duplicate code.

suresh-thelkar replied:
This was added to take care in case if an Admin does not update sched_host/sched_port/sched_priv/sched_log of scheduler object and if the stat goes to the server instance where scheduler is not running. 
As per our discussion I felt that if we mandate the Admin to set up the following then this change is not required and hence removed it.

"Admin has to setup the following in case of Multi-Server setup even if all servers run in the same machine"
     qmgr -c "s sc sched_host=spark1,sched_port=15004"
    qmgr -c "s sched sched_log=/var/spool/pbs/sched_logs"
    qmgr -c "s sched sched_priv=/var/spool/pbs/sched_priv"

nithinj commented on src/lib/Libifl/pbsD_selectj.c:
```diff
@@ -263,15 +268,27 @@ __pbs_selstat(int c, struct attropl *attrib, struct attrl *rattrib, char *extend
-	    PBS_BATCH_SelectJobs, MGR_OBJ_JOB, MGR_CMD_NONE, attrib))
+	for (i = 0; !(attrs_verified) && i < num_cfg_svrs; i++) {
+		/* first verify the attributes, if verification is enabled */
+		if ((pbs_verify_attributes(svr_connections[i].sd, PBS_BATCH_SelectJobs,
```
why this change? PBSD_status_random gets you an already connected server

suresh-thelkar replied:
As discussed PBSD_status_random alone is not sufficient. I have written a wrapper for pbs_verify_attributes and used wherever needed. Please check my latest commit.

nithinj commented on src/lib/Libifl/pbsD_selectj.c:
```diff
@@ -198,8 +198,11 @@ __pbs_selectjob(int c, struct attropl *attrib, char *extend)
 
 	for (i = 0; i < get_num_servers(); i++) {
-		if (svr_connections[i].state != SVR_CONN_STATE_CONNECTED)
+		if (svr_connections[i].state != SVR_CONN_STATE_CONNECTED) {
```
Handling in the application is better. We do not have to inject application-specific cases within IFL layer.

nithinj commented on src/lib/Libifl/int_status.c:
```diff
@@ -546,14 +548,26 @@ PBSD_status_aggregate(int c, int cmd, char *id, struct attrl *attrib, char *exte
 
-	for (i = 0; i < get_num_servers(); i++) {
-		if (svr_connections[i].state != SVR_CONN_STATE_CONNECTED)
+	for (i = 0; i <num_cfg_svrs; i++) {
```
Please remove this check. Keep IFL generic and handle these cases in the scheduler

nithinj commented on src/lib/Libifl/int_status.c:
```diff
@@ -546,14 +548,26 @@ PBSD_status_aggregate(int c, int cmd, char *id, struct attrl *attrib, char *exte
-		parent_object, MGR_CMD_NONE, (struct attropl *) attrib)))
+	for (i = 0; !(attrs_verified) && i < num_cfg_svrs; i++) {
+		/* first verify the attributes, if verification is enabled */
+		if ((pbs_verify_attributes(svr_connections[i].sd, cmd,
```
random_srv_conn will get you a connected server. This loop is unnecessary.

suresh-thelkar replied:
This is needed as per our discussion. Please look into my latest commit.

nithinj commented on src/lib/Libifl/int_manager.c:
```diff
@@ -85,32 +92,57 @@ PBSD_manager(int c, int function, int command, int objtype, char *objname, struc
-	if (pbs_client_thread_lock_connection(c) != 0)
+	for (i = 0; !(attrs_verified) && i < num_cfg_svrs; i++) {
+		/* now verify the attributes, if verification is enabled */
+		if ((pbs_verify_attributes(svr_connections[i].sd, function, objtype, command, aoplp)))
```
Use random_srv_conn instead of looping here.

nithinj commented on src/lib/Libifl/int_manager.c:
```diff
@@ -85,32 +92,57 @@ PBSD_manager(int c, int function, int command, int objtype, char *objname, struc
-	}
+	for (i = 0; i < num_cfg_svrs; i++) {
+		if (svr_connections[i].state != SVR_CONN_STATE_CONNECTED) {
+			if (svr_connections[i].from_sched)
```
No application-specific check, please.

nithinj commented:
@suresh-thelkar ,
I've raised https://github.com/subhasisb/openpbs/pull/136 . qmgr will not proceed if one of the connection is broken, and print the appropriate message to the console with server name.

suresh-thelkar commented:
> @suresh-thelkar ,
> I've raised #136 . qmgr will not proceed if one of the connection is broken, and print the appropriate message to the console with server name.

As we discussed both of our changes are needed. One of us can rebase with others changes after the merge of one of these PRs so that we get both of the changes.

suresh-thelkar commented:
@nithinj, I have implemented all your comments as we discussed. Please take a look and provide your further feedback.

nithinj commented on src/lib/Libecl/ecl_verify_object_name.c:
```diff
@@ -163,3 +164,55 @@ pbs_verify_object_name(int type, char *name)
+	if (!svr_connections)
+		return (pbs_errno = PBSE_NOSERVER);
+
+	for (i = 0; !(attrs_verified) && i < num_cfg_svrs; i++) {
```
We should be bailing out if one of them succeeds or fails, and the failure code is not connected.

suresh-thelkar replied:
Done

agrawalravi90 commented:
> qmgr/pbs_manager request should go to all servers automatically.
Even if one of the configured servers is not available qmgr update should fail. But the stats queries should be successful when queried through external clients. But when they are called from scheduler the operation should fail.

do ALL qmgr requests need to go to all servers? only the requests for shared objects need to go to all right? if nodes are divided then a qmgr request for a node only needs to go to one server right?

suresh-thelkar commented:
> > qmgr/pbs_manager request should go to all servers automatically.
> > Even if one of the configured servers is not available qmgr update should fail. But the stats queries should be successful when queried through external clients. But when they are called from scheduler the operation should fail.
> 
> do ALL qmgr requests need to go to all servers? only the requests for shared objects need to go to all right? if nodes are divided then a qmgr request for a node only needs to go to one server right?

When admin creates nodes he has to use PBS_SERVER_INSTANCES and create them. For example 
PBS_SERVER_INSTANCES=spark qmgr -c "c n stblr3"
Then the updates will only go to the respective server.

When it comes to multi-node jobs right now as we know @nithinj  is working on couple of the approaches. One of them needs updates to all Servers. Also there happened a good of amount of discussion on the same between myself, @hirenvadalia  and @nithinj in our team's discussion. Can you please take look for more details.

suresh-thelkar commented on src/scheduler/fifo.c:
```diff
@@ -2909,29 +2906,13 @@ update_svr_schedobj(int connector, int cmd, int alarm_time)
-	patt->value = port_str;
-	patt->next = patt + 1;
-	patt++;
 	patt->name = ATTR_version;
```
One reason to think of having ATTR_version is to support Server/Scheduler at different versions of PBS. If we no longer support that and there is no other historical reason I can go ahead and delete the code where Scheduler updates this to Server.

Also alarm_time is deprecated and as we have substitution for that with ATTR_sched_cycle_len that can be set through qmgr I can clean up this code as well. We can make these changes in our rapid branch without any problems. 

@subhasisb, can you please let me know is it ok to clean up these code and remove these attributes in the mainline as well. Also just wanted to know how long we keep the deprecated code i.e. is it two or three releases.

hirenvadalia replied:
@suresh-thelkar, I am already removing this code in mainline...

hirenvadalia replied:
See this: https://openpbs.atlassian.net/wiki/spaces/PD/pages/2047344647/Make+scheduler+to+connect+server+and+keep+that+connection+persistent

suresh-thelkar replied:
@hirenvadalia, As per your document it seems you are still sending the scheduler's version to Server. Are you planning to remove that as well ?

hirenvadalia replied:
@suresh-thelkar, if we don't need then yes I can remove, but I also have same question as Ravi... why we are sending version to server?

suresh-thelkar replied:
@hirenvadalia, I have mentioned earlier one of the reasons why this can be used. I am also eager to know the reason for sending scheduler's version to server. I think @subhasisb can help us in getting the historical reason for this. @subhasisb, can you please let us know your thoughts on this.

hirenvadalia replied:
AFAIK as of now we don't support different version of daemon in same cluster...

agrawalravi90 replied:
> One reason to think of having ATTR_version is to support Server/Scheduler at different versions of PBS

Scheduler gets installed as part of the same rpm as the server, so a customer site should never have a scheduler which has a different version from the server, right?

suresh-thelkar replied:
One reason that I mentioned is what I can think of. May be there is a historical reason behind this. I haven sent an email to Bhroam regarding this. 
I also had chat with @subhasisb regarding this and Subhasis suggested me the following. 

We are keeping it now in our branch as it does not affect our performance since Scheduler updates this attribute to each server only once. Once we know the reason behind this attribute and we decide to remove it, then we can straightaway clean up the code in mainline and sync it with our branch that time.

agrawalravi90 replied:
Ya, it's not a performance concern, it's just that I saw an opportunity to remove this function altogether. We don't have to hold this PR for this, but if we do get confirmation from Bhroam that this can be removed before all the other comments are addressed then please do address this comment.

suresh-thelkar replied:
> Ya, it's not a performance concern, it's just that I saw an opportunity to remove this function altogether. We don't have to hold this PR for this, but if we do get confirmation from Bhroam that this can be removed before all the other comments are addressed then please do address this comment.

Sure. If we get a clarity on this and I can implement in this PR itself.

suresh-thelkar commented on src/server/sched_func.c:
```diff
@@ -616,9 +616,6 @@ action_sched_priv(attribute *pattr, void *pobj, int actmode)
 
 	psched = (pbs_sched *) pobj;
 
-	if (pobj == dflt_scheduler)
```
Myself and @subhasisb had a chat earlier and we thought that we should not have any differentiation with respect to default scheduler and MultiSched. And as we know we had to allow Admin to change sched_host/sched_port as all other servers running in different PBS complexes should also need to connect to the same scheduler. On the same lines we should also allow sched_priv/sched_logs to be changed by the Admin if they want a non default location. 

There is another reason for removing this restriction. In case of Multi-Server as I have explained in the PR description we are now recommending or mandating admin to set sched_host, sched_port, sched_priv and sched_log. If we do not mandate this then qmgr -c "p sched" output differs from server to server. 

For example if qmgr request goes to server-1 we can get the following output
```
[root@stblr3 pbspro]# qmgr -c "p sched sched_host,sched_port,sched_priv,sched_log"
#
# Create and define scheduler default
#
create sched default
set sched sched_log = /var/spool/pbs/sched_logs
set sched sched_priv = /var/spool/pbs/sched_priv
set sched sched_port = 15004
set sched sched_host = stblr3
```

When it goes to server-2 we can get the following output
```
[root@stblr3 pbspro]# qmgr -c "p sched sched_host,sched_port,sched_priv,sched_log"
#
# Create and define scheduler default
#
create sched default
set sched sched_log = /var/spool/pbs_2/sched_logs
set sched sched_priv = /var/spool/pbs_2/sched_priv
set sched sched_port = 15004
set sched sched_host = stblr3

```


Even without this restriction default scheduler's priv and log are populated to have the fixed location which is by default derived from its PBS_HOME in set_sched_defaults(). So I don't think we need to change PTL,pbs_snapshot etc.

agrawalravi90 replied:
"There is another reason for removing this restriction. In case of Multi-Server as I have explained in the PR description we are now recommending or mandating admin to set sched_host, sched_port, sched_priv and sched_log. If we do not mandate this then qmgr -c "p sched" output differs from server to server."

I understand the need for sched_port/host to be customizable, but I'm still not convinced that we need to make sched log and priv customizable. Why would qmgr -c "p sched" differ from server to server today? Is it because we don't print the default sched_log and sched_priv values of default sched? If yes, then that can be fixed very easily. 

I know for a fact that pbs_snapshot and simulator will need some changes, maybe not much. But regardless of that, we should only make this customizable if needed, I personally feel that we should have never made sched_log and sched_priv customizable for any multisched, we could have just chosen a naming scheme internally for the admin and saved ourselves the pain of job sort formula bugs that this caused. Anyways, has any admin asked for this to be customizable? If not, then can you please explain more why this is necessary for multi-server to work?

suresh-thelkar replied:
> I understand the need for sched_port/host to be customizable, but I'm still not convinced that we need to make sched log and priv customizable. Why would qmgr -c "p sched" differ from server to server today? Is it because we don't print the default sched_log and sched_priv values of default sched? If yes, then that can be fixed very easily, right?

As far as I remember correctly the requirement of having the need to configure sched_priv/sched_logs has come from some customers/PMs. There was a plan to move job_sort_formula from server to sched object and for some reasons it is not prioritised as we have a work around for that. I don't know the current state of it though.

Anyways, Coming to Multi-Server following is my explanation. 

It all depends on how we are setting up the Multi-Server. For example we can have couple different ways to set up 3 servers as follows.

1. Server-1 in PBS Complex -1 with all other daemons running on the same machine M1 including the scheduler.
    Server-2 in PBS Complex -2 with only MOM running in a different machine M2
    Server-3 in PBS Complex -3 with only MOM running in a different machine M3
2. Server-1 in PBS Complex -1 with all other daemons running on the same machine M1 including the scheduler.
     Server-2 with only MOM running in machine M1 itself
     Server-3 with only MOM running in machine M1 itself

When we do qmgr -c "p sched" default values are printed even for default scheduler which is not a problem. 
When it comes to configuration-1 given above qmgr -c "p sched" output does not differ from server to server.
But when it comes to configuration-2, qmgr -c "p sched" output differs from server to server. This is because on the same machine M1 we have setup PBS Servers with different PBS_HOME. In fact the output that I have given earlier(https://github.com/subhasisb/openpbs/pull/135#discussion_r482692364) is with respect to this configuration. Please observe the difference in sched_priv/sched_log outputs.

In order to not see the different outputs in this configuration our recommendation is to mandate Admin to set sched_host, sched_port, sched_priv, sched_log in case of Multi-Server setup. If this is followed then all servers get the update and store these attributes in database because our qmgr requests are broadcasted after which they get the uniform/same output when they run qmgr -c "p sched"

agrawalravi90 replied:
"This is because on the same machine M1 we have setup PBS Servers with different PBS_HOME. In fact the output that I have given earlier(#135 (comment)) is with respect to this configuration. Please observe the difference in sched_priv/sched_log outputs."

ok, i see it now .. interesting .. I still think that this can be solved without making this restriction, but I guess for now this is the simplest way of fixing it.

agrawalravi90 commented:
> Server should stop contacting Scheduler when scheduling is off. This helps in not logging the following message to server logs continuously thus by filling up all the system when scheduler is gone down.

This should be taken care of by Hiren's connection reversal PR right?

suresh-thelkar commented:
> > Server should stop contacting Scheduler when scheduling is off. This helps in not logging the following message to server logs continuously thus by filling up all the system when scheduler is gone down.
> 
> This should be taken care of by Hiren's connection reversal PR right?

Yes @agrawalravi90. But It might take some more time to actually get these things in because the design discussions are still going on for the master branch. So @subhasisb asked me to do this for the time being in our rapid branch so that it does not accidentally fill up the system's disk if scheduler is shutdown and servers are still up.

suresh-thelkar commented:
@agrawalravi90, I have addressed all your comments. I also added code changes for the following bugs + new change requests.
1. qstat without using PBS_SERVER_INSTANCES is not printing the partial data if one of the servers is down. However if we use PBS_SERVER_INSTANCES to point to the server/s that are active then we get the partial data. This causes problems especially if we have more number of servers and user does not know which server is down.
The ideal thing here is qstat should give the partial data and at the same time it should also notify the user that a particular server instance is down so that he is aware that he is getting partial data.
2. There was bug in the connect_to_servers() function. In case of multiple servers when last server is down this function used to return -1 and the clients think that no servers are active or connected.
   Modified the function so as to send the first connected fd if we have at least one server is connected.
3. Scheduler code clean up to remove deprecated attributes
4. Display Formatting changes to qstat.c/qmgr.c in case of Multi-Server. 

Please look into my last commit for changes.

suresh-thelkar commented:
I have also attached in detail logs for all the scenarios. 
[Misc_Sched_ERs.log](https://github.com/subhasisb/openpbs/files/5174351/Misc_Sched_ERs.log)

suresh-thelkar commented:
@agrawalravi90, I have also done the code clean-up to remove the function update_svr_schedobj and the corresponding code and added as a separate commit. Please take a look.

agrawalravi90 commented on src/lib/Libifl/int_status.c:
```diff
@@ -546,14 +548,15 @@ PBSD_status_aggregate(int c, int cmd, char *id, struct attrl *attrib, char *exte
-	for (i = 0; i < get_num_servers(); i++) {
-		if (svr_connections[i].state != SVR_CONN_STATE_CONNECTED)
+	for (i = 0; i <num_cfg_svrs; i++) {
+		if (svr_connections[i].state != SVR_CONN_STATE_CONNECTED) {
```
so, if an admin has a pbs.conf which has 3 servers listed, but he has only 2 servers up for whatever reason, will users keep seeing PBSE_NOSERVER error for every client request?

agrawalravi90 commented on src/lib/Libifl/int_status.c:
```diff
@@ -707,7 +709,7 @@ PBSD_status_get(int c, int svr_index)
 			rbsp->last = bsp;
 
-			if (from_sched) {
+			if ((svr_index != -1) && (svr_conns[svr_index].from_sched)) {
```
why did you add from_sched? checking secondary_sd was not the correct way to check for sched?

suresh-thelkar commented on src/scheduler/pbs_sched.c:
```diff
@@ -1560,22 +1550,14 @@ schedule_wrapper(fd_set *read_fdset, int opt_no_restart)
-
-				if (num_svrs_updated < num_cfg_svrs) {
-					/* update sched object attributes on server */
-					if (update_svr_schedobj(ifl_sock, cmd, alarm_time) == 0) {
```
As part of SCH_SCHEDULE_FIRST this function gets called and hence I don't think this is needed here.

suresh-thelkar commented on src/cmds/qstat.c:
```diff
@@ -3093,9 +3121,12 @@ qstat -B [-f] [-F format] [-D delim] [ server_name... ]\n";
 					break;
 				}
+
+				show_svr_inst_fail();
```
For jobs, queues etc we have labels, I used to call this function once per each label. This is because only one type of objects are queried at a time. Anyways now I am calling this at the start of the loop once which takes care of this.

suresh-thelkar commented:
@agrawalravi90 , I have implemented your review comments. Please check.

agrawalravi90 reviewed (approved):
LGTM
